### PR TITLE
Verify assumptions in SCION.DecodeFromBytes

### PR DESCRIPTION
**Pull request by mlimbeck**
Co-authored by: @Dspil

### CODE REVIEW

File: pkg/slayers/scion_spec.gobra [hunk 1]
```diff
@@ -431,6 +431,21 @@ func (s *SCION) EqAbsHeader(ub []byte) bool {
 		scion.Base{metaHdr, scion.NumInfoFields(seg1, seg2, seg3), seg1 + seg2 + seg3}
 }
 
+ghost
+opaque
+pure
+requires acc(s.Mem(ub), _)
+decreases
+func (s *SCION) ValidSCIONInitSpec(ub []byte) bool {
```
Comment (jcp19) [inline pkg/slayers/scion_spec.gobra:439]: What is this meant to mean? What does `Init` mean here? Can you please add a comment explaining the purpose of this function

File: router/dataplane.go [hunk 2]
```diff
@@ -1955,10 +1957,20 @@ func (p *scionPacketProcessor) parsePath( /*@ ghost ub []byte @*/ ) (respr proce
 		// TODO(lukedirtwalker) parameter problem invalid path?
 		return processResult{}, err
 	}
-	// @ TemporaryAssumeForIO(slayers.ValidPktMetaHdr(ub))
+	// (VerifiedSCION) This assumption cannot be proven at this point.
+	// There might be a check is missing.
+	// @ TemporaryAssumeForIO(p.path.CurrInfMatchesCurrHF(ubPath))
```
Comment (jcp19) [inline router/dataplane.go:1962]: As a reminder for the future, please report this in `scionproto/scion`

File: pkg/slayers/path/epic/epic.go [hunk 3]
```diff
@@ -148,8 +148,8 @@ func (p *Path) DecodeFromBytes(b []byte) (r error) {
 		return serrors.New("EPIC Path raw too short", "expected", int(MetadataLen), "actual", int(len(b)))
 	}
 	//@ unfold p.NonInitMem()
-	//@ slices.SplitByIndex_Bytes(b, 0, len(b), PktIDLen, R40)
-	//@ preserves acc(slices.AbsSlice_Bytes(b, 0, PktIDLen), R40)
+	//@ slices.SplitByIndex_Bytes(b, 0, len(b), PktIDLen, R42)
+	//@ preserves acc(slices.AbsSlice_Bytes(b, 0, PktIDLen), R42)
```
[the remark this answers is not in the document]
Comment (mlimbeck) [reply pkg/slayers/path/epic/epic.go]: No performance difference at all. I removed all `outline` blocks and simplified the  `unfolding/folding`.

File: pkg/slayers/path/io_msgterm_spec.gobra [hunk 4]
```diff
@@ -28,8 +31,22 @@ decreases
...
+ensures len(res) == MacLen
+ensures forall i int :: { mac[i] } 0 <= i && i < MacLen ==> res[i] == mac[i]
 decreases
-pure func AbsMac(mac [MacLen]byte) (io.IO_msgterm)
+pure func macSequence(mac [MacLen]byte) (res seq[byte])
+
+ghost
+requires len(mac) == MacLen
+decreases
+pure func absMac_Internal(mac seq[byte]) (io.IO_msgterm)
```
[the remark this answers is not in the document]
Comment (mlimbeck) [reply pkg/slayers/path/io_msgterm_spec.gobra:42]: We need these functions for `AbsMac_Lemma()` which is used to prove `DecodeFromBytes` for `Hopfields`.

File: pkg/slayers/path/path.go [hunk 5]
```diff
@@ -81,12 +81,19 @@ type Path interface {
...
 	//@ ensures   err != nil ==> NonInitMem()
+	//@ ensures   err == nil ==> DecodeFromBytesSpec(b, err)
 	//@ decreases
 	DecodeFromBytes(b []byte) (err error)
+	//@ ghost
+	//@ pure
+	//@ requires  Mem(b)
+	//@ requires  acc(sl.AbsSlice_Bytes(b, 0, len(b)), R42)
+	//@ decreases
+	//@ DecodeFromBytesSpec(b []byte, err error) (res bool)
```
[the remark this answers is not in the document]
Comment (mlimbeck) [reply pkg/slayers/path/path.go]: No, there are not related. I renamed it.

File: pkg/slayers/scion_spec.gobra (hunk 1 above)
Comment (mlimbeck) [reply pkg/slayers/scion_spec.gobra]: I added I comment. Can you check if it is clear now?

File: pkg/slayers/scion_spec.gobra [hunk 6]
```diff
@@ -431,6 +431,21 @@ func (s *SCION) EqAbsHeader(ub []byte) bool {
...
+	return unfolding acc(s.Mem(ub), _) in
+		let low := CmnHdrLen+s.AddrHdrLenSpecInternal() in
+		let high := s.HdrLen*LineLen in
+		typeOf(s.Path) == (*scion.Raw)			&&
```
[the remark this answers is not in the document]
Comment (mlimbeck) [reply pkg/slayers/scion_spec.gobra]: Can you check if its better now?

File: pkg/slayers/path/scion/raw_spec.gobra [hunk 7]
```diff
@@ -879,4 +1054,50 @@ func XoverRightSeg(raw []byte, currInfIdx int, currHfIdx int, seg1Len int, seg2L
...
+requires  len(hopFields) == segLen - currHFIdx
+requires  offset + path.HopLen * segLen <= len(raw)
+requires  acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R56)
+decreases
+pure func hopFieldsBytePositions(
+	raw []byte,
+	offset int,
+	currHFIdx int,
+	segLen int,
+	hopFields seq[io.IO_HF]) bool {
```
[the remark this answers is not in the document]
Comment (mlimbeck) [reply pkg/slayers/path/scion/raw_spec.gobra]: Can you check that the line isn't to long now?

File: pkg/slayers/path/scion/raw.go [hunk 8]
```diff
@@ -480,7 +507,12 @@ func (s *Raw) GetCurrentHopField( /*@ ghost ubuf []byte @*/ ) (res path.HopField
 	//@ assume 0 <= idx
 	//@ fold acc(s.Base.Mem(), R10)
 	//@ fold acc(s.Mem(ubuf), R9)
-	return s.GetHopField(idx /*@, ubuf @*/)
+	// (VeifiedSCION) This return is substituted with an assignment so that
+	// we can call a lemma afterwards and satisfy a postcondition.
+	res, r = s.GetHopField(idx /*@, ubuf @*/)
+	//@ assert r == nil ==> reveal s.CorrectlyDecodedHf(ubuf, res) ==
+	//@ 	reveal s.CorrectlyDecodedHfWithIdx(ubuf, idx, res)
```
[the remark this answers is not in the document]
Comment (mlimbeck) [reply pkg/slayers/path/scion/raw.go]: Can you check if I picked the right trigger?

File: pkg/slayers/path/scion/raw_spec.gobra [hunk 9]
```diff
@@ -125,6 +135,22 @@ pure func (s *Raw) ValidCurrIdxs(ghost ub []byte) bool {
...
+		s.Base.InfsMatchHfs()
+}
+
+ghost
+requires acc(s.Mem(ub), _)
+decreases
+pure func (s *Raw) CurrInfMatchesCurrHF(ghost ub []byte) bool {
+	return unfolding acc(s.Mem(ub), _) in
+		s.Base.GetBase().CurrInfMatchesCurrHF()
+}
```
Comment (jcp19) [inline pkg/slayers/path/scion/raw_spec.gobra:152]: As a general comment, this is fine but I plan to deprecate all these wrappers. First, they hide the fact that these functions only mention fields in the Base. Second, to know the definitions, we need to follow a few layers of wrappers, which is often frustrating. Instead, in #347, I will drop all of these, introduce a GetBase() function for `*Raw` and the other types, and write all specs like `s.GetBase(ub).CurrInfMatchesCurrHF()`. While this is longer, it addresses the two points I mentioned

File: pkg/slayers/scion_spec.gobra [hunk 10]
```diff
@@ -413,24 +413,41 @@ func (s *SCION) EqAbsHeader(ub []byte) bool {
 		let low := CmnHdrLen+s.AddrHdrLenSpecInternal() in
 		let high := s.HdrLen*LineLen in
 		GetAddressOffset(ub) == low &&
-		GetLength(ub) == int(high) &&
+		GetLength(ub) == int(high) 	&&
```
Comment (jcp19) [inline pkg/slayers/scion_spec.gobra:416]:
```suggestion
		GetLength(ub) == int(high)        &&
```

File: pkg/slayers/scion_spec.gobra [hunk 11]
```diff
@@ -413,24 +413,41 @@ func (s *SCION) EqAbsHeader(ub []byte) bool {
...
+opaque
+pure
+requires acc(s.Mem(ub), _)
+decreases
+func (s *SCION) ValidScionInitSpec(ub []byte) bool {
+	return unfolding acc(s.Mem(ub), _) 					in
+		let low := CmnHdrLen+s.AddrHdrLenSpecInternal() in
+		let high := s.HdrLen*LineLen 					in
+		typeOf(s.Path) == (*scion.Raw) &&
+		unfolding acc(s.Path.Mem(ub[low:high]), _) 		in
```
Comment (jcp19) [inline pkg/slayers/scion_spec.gobra:446]: please don't use tabs before the `in`, it looks very weird on github and it is editor dependent

File: router/dataplane.go [hunk 12]
```diff
@@ -1743,8 +1743,8 @@ func (p *scionPacketProcessor) processIntraBFD(data []byte) (res error) {
 // contracts for IO-spec
 // @ requires  p.d.DpAgreesWithSpec(dp)
 // @ requires  dp.Valid()
-// @ requires  (typeOf(p.scionLayer.GetPath(ub)) == *scion.Raw) ==> slayers.ValidPktMetaHdr(ub)
-// @ requires  (typeOf(p.scionLayer.GetPath(ub)) == *scion.Raw) ==> p.scionLayer.EqAbsHeader(ub)
+// @ requires  (typeOf(p.scionLayer.GetPath(ub)) == *scion.Raw) ==>
+// @	p.scionLayer.EqAbsHeader(ub) && p.scionLayer.ValidScionInitSpec(ub)
```
Comment (jcp19) [inline router/dataplane.go:1747]:
```suggestion
// @ 	p.scionLayer.EqAbsHeader(ub) && p.scionLayer.ValidScionInitSpec(ub)
```

Comment (jcp19) [reply router/dataplane.go:1747]: all lines should have a space after the `@`

File: pkg/slayers/scion_spec.gobra [hunk 13]
```diff
@@ -413,24 +413,41 @@ func (s *SCION) EqAbsHeader(ub []byte) bool {
...
-		let seg3 := int(metaHdr.SegLen[2])       in
+		let seg1 := int(metaHdr.SegLen[0]) in
+		let seg2 := int(metaHdr.SegLen[1]) in
+		let seg3 := int(metaHdr.SegLen[2]) in
 		s.Path.(*scion.Raw).Base.GetBase() ==
 		scion.Base{metaHdr, scion.NumInfoFields(seg1, seg2, seg3), seg1 + seg2 + seg3}
 }
 
+// Contains all specifications that hold
+// after a scion packet is decoded by DecodeFromBytes().
```
Comment (jcp19) [inline pkg/slayers/scion_spec.gobra:435]:
```suggestion
// Describes a SCION packet that was successfully decoded by `DecodeFromBytes`.
```

**Review by jcp19 — APPROVED** [review]
Minor comments about formatting

File: pkg/slayers/path/io_msgterm_spec.gobra [hunk 14]
```diff
@@ -44,3 +61,24 @@ ensures  len(res) == MacLen
 ensures  forall i int :: { res[i] } 0 <= i && i < MacLen ==> mac[i] == res[i]
 decreases
 pure func FromSliceToMacArray(mac []byte) (res [MacLen]byte)
```
[the remark this answers is not in the document]
Comment (jcp19) [reply pkg/slayers/path/io_msgterm_spec.gobra:63]: This was just merged btw

File: pkg/slayers/path/io_msgterm_spec.gobra [hunk 15]
```diff
@@ -44,3 +61,24 @@ ensures  len(res) == MacLen
...
+
+
+ghost
+requires len(mac1) == MacLen
+requires forall i int :: { &mac1[i] } 0 <= i && i < MacLen ==> acc(&mac1[i], R50)
+requires forall i int :: { &mac1[i] } 0 <= i && i < MacLen ==> mac1[i] == mac2[i]
+ensures  forall i int :: { &mac1[i] } 0 <= i && i < MacLen ==> acc(&mac1[i], R50)
+ensures  AbsMac(FromSliceToMacArray(mac1)) == AbsMac(mac2)
+decreases
+func AbsMac_Lemma(mac1 []byte, mac2 [MacLen]byte) {
```
Comment (jcp19) [inline pkg/slayers/path/io_msgterm_spec.gobra:73]: Please rename this to `EqualBytesImplyEqualMac`, let's not use underscores here.

Also, the postcondition here should be `FromSliceToMacArray(mac1) == mac2`, which is stronger than the current post, and I think that we can simplify the proof

File: pkg/slayers/path/io_msgterm_spec.gobra (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcp19) [reply pkg/slayers/path/io_msgterm_spec.gobra:42]: I don't see why. We should be able to prove that `FromSliceToMacArray(mac1) == mac2`, especially after adding a body to `FromSliceToMacArray`. From this, the proof of `AbsMac_Lemma` follows trivially. I don't see how introducing `absMac_Internal` helps in any way here.

**Pull request merged**